pkg/controller/sparkapplication/controller.go

```diff
@@ -289,397 +298,464 @@ func (c *Controller) processNextItem() bool {
+//|    +------------+                           |                               |                                      |
+//|                                             |                               |                                      |
+//|                                             +-------------------------------+                                      |
+//|                                                                                                                    |
+//+--------------------------------------------------------------------------------------------------------------------+
 
-	c.recordDriverEvent(app, update.podPhase, update.podName)
-
-	// The application state is solely based on the driver pod phase once the application is successfully
-	// submitted and the driver pod is created.
-	appState := driverPodPhaseToApplicationState(update.podPhase)
-	if isAppTerminated(appState) {
-		c.recorder.Eventf(
-			app,
-			apiv1.EventTypeNormal,
-			"SparkApplicationTerminated",
-			"SparkApplication %s terminated with state: %v",
-			update.appName,
-			appState)
+func (c *Controller) syncSparkApplication(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return fmt.Errorf("failed to get the namespace and name from key %s: %v", key, err)
 	}
+	app, err := c.getSparkApplication(namespace, name)
+	if err != nil {
+		return err
+	}
+	appToUpdate := app.DeepCopy()
 
-	return c.updateSparkApplicationStatusWithRetries(app, func(status *v1alpha1.SparkApplicationStatus) {
-		status.DriverInfo.PodName = update.podName
-		if update.nodeName != "" {
-			if nodeIP := c.getNodeExternalIP(update.nodeName); nodeIP != "" {
-				status.DriverInfo.WebUIAddress = fmt.Sprintf("%s:%d", nodeIP,
-					status.DriverInfo.WebUIPort)
+	// CRD Delete requested.
+	if !appToUpdate.GetDeletionTimestamp().IsZero() {
+		appToUpdate = c.handleSparkApplicationDeletion(appToUpdate)
+	} else {
+		// Take action based on AppState.
+		switch appToUpdate.Status.AppState.State {
+		case v1alpha1.NewState:
+			c.recordSparkApplicationEvent(appToUpdate)
+			appToUpdate.Status.SubmissionAttempts = 0
+			glog.Infof("Creating Submission for SparkApplication: %s", key)
+			appToUpdate = c.submitSparkApplication(appToUpdate)
+		case v1alpha1.SucceedingState:
+			if shouldRetry(appToUpdate) {
+				if err := c.deleteSparkResources(appToUpdate, true); err != nil {
+					glog.Errorf("Failed to delete the driver pod and UI service for deleted SparkApplication %s/%s. Error: %v",
+						appToUpdate.Namespace, appToUpdate.Name, err)
+					return err
+				}
+				appToUpdate.Status.AppState.State = v1alpha1.PendingRetryState
+			} else {
+				// App will never be retried. Move to Terminal Completed State.
+				appToUpdate.Status.AppState.State = v1alpha1.CompletedState
+				c.recordSparkApplicationEvent(appToUpdate)
+			}
+		case v1alpha1.FailingState:
+			if !shouldRetry(appToUpdate) {
```

Comment:
Nit: `shouldRetry(appToUpdate)` is used above, so to be consistent, the if block should also check `shouldRetry(appToUpdate)` here.